yew/src/virtual_dom/vlist.rs

```diff
@@ -78,100 +85,148 @@ impl VDiff for VList {
-                        previous_sibling =
-                            left.apply(scope, parent, previous_sibling.as_ref(), None);
-                    }
+
+            let right_key = rights
+                .get(current_right_idx)
+                .and_then(|vnode| vnode.key().as_ref());
+            let (right_idx, reused) = match (left.key(), right_key) {
+                (Some(left_key), Some(right_key)) if left_key == right_key => {
+                    current_right_idx += 1;
+                    (Some(current_right_idx - 1), false)
                 }
-            } else {
-                match rights.next() {
-                    Some(right) => {
-                        previous_sibling =
-                            left.apply(scope, parent, previous_sibling.as_ref(), Some(right));
+                (Some(left_key), _) => match rights_lookup.get(left_key) {
+                    Some(id) if *id > current_right_idx => {
+                        // We found the existing node with the same key, but its
+                        // order has been changed, so we must move it to its new
+                        // location if we want to reuse the node.
+                        (Some(*id), true)
+                    }
+                    _ if left_surplus > 0 => {
+                        left_surplus = left_surplus.saturating_sub(1);
+                        (None, true)
                     }
-                    None => {
-                        previous_sibling =
-                            left.apply(scope, parent, previous_sibling.as_ref(), None);
+                    _ => {
+                        current_right_idx += 1;
+                        (Some(current_right_idx - 1), false)
                     }
+                },
+                (None, _) => {
+                    current_right_idx += 1;
+                    (Some(current_right_idx - 1), false)
+                }
+            };
+
+            let right = right_idx.and_then(|right_idx| {
+                rights.get_mut(right_idx).map(|right| {
+                    let dummy_vnode = VNode::VText(VText::new(String::default()));
+                    core::mem::replace(right, dummy_vnode)
+                })
+            });
+
+            if reused {
+                if let Some(right) = &right {
+                    // Move the ancestor node.
+                    super::insert_node(
+                        &right.reference().expect("there must be a reference"),
+                        parent,
+                        &node_position,
+                    );
+                } else {
+                    // If there isn't an ancestor, then the new node will be
+                    // inserted at the correct position so there is nothing to
+                    // do here.
                 }
             }
+
+            node_position = match left.apply(parent_scope, parent, node_position, right) {
+                Some(node) => VDiffNodePosition::After(node),
```

Comment:
It would be awesome if we could refactor to avoid using this variant. We already have the full list of `rights` so we should be able to exclusively use `VDiffNodePosition::Before`. That way we can avoid the extra DOM api call to get the next sibling